components/ContactList.js

```diff
@@ -27,18 +34,60 @@ import ContactItem from "../components/ContactItem";
 // .....
 
 export default function ContactList() {
+  // Hooks
   const dispatch = useDispatch();
   const contacts = useSelector(getContacts);
+  const favorites = useSelector(getFavorites);
+  const [showFavorites, setShowFavorites] = useState(false);
 
   useEffect(() => {
     dispatch(fetchContacts());
   }, []);
 
+  // JSX
+  const contactsJSX = contacts.map((contact, index, array) => {
+    const currentLetter = contact.lastName[0].toUpperCase();
+    const favorite = favorites.some((favorite) => favorite.id === contact.id);
+    if (!index) {
+      // Provides header for first contact item
+      return (
+        <Fragment key={index}>
+          <p>{currentLetter}</p>
+          <ContactItem key={index} {...contact} favorite={favorite} />
+        </Fragment>
+      );
+    } else if (index < array.length - 1) {
+      const nextLetter = array[index + 1].lastName[0].toUpperCase();
+      if (currentLetter !== nextLetter) {
+        // Provides header for all other contact items
+        return (
+          <Fragment key={index}>
+            <ContactItem key={index} {...contact} favorite={favorite} />
+            <p>{nextLetter}</p>
+          </Fragment>
+        )
+      } else {
+        return (
+          <ContactItem key={index} {...contact} favorite={favorite} />
+        )
+      }
+    }
+  });
```

Comment:
This is a really unique approach. I think it works pretty well!
Pros:
- only requires one map, so if you had thousands of contacts it would still be pretty performant
Cons:
- it's not exactly the most readable solution imo, but your comments help here a bit

Reply:
Yeah. I agree it is not the most readable solution and in hindsight it would have been probably better to put this inside its own component.